backend/api/Services/DocumentService.cs

```diff
@@ -129,29 +134,88 @@ public async Task<DocumentUploadResponse> UploadActivityDocumentAsync(long activ
             this.User.ThrowIfNotAuthorized(Permissions.DocumentEdit);
 
-            // update metadata of document
-            IList<Task<ExternalResult<DocumentMetadata>>> metadataTasks = new List<Task<ExternalResult<DocumentMetadata>>>();
-            foreach (var metadata in updateRequest.DocumentMetadata ?? new List<DocumentMetadataUpdateModel>())
+            // update the pims document status
+            PimsDocument existingDocument = documentRepository.Get(documentId);
+            if (existingDocument == null)
             {
-                metadataTasks.Add(documentStorageRepository.UpdateDocumentMetadataAsync(updateRequest.MayanDocumentId, metadata.Id, metadata.Value));
+                throw new BadRequestException("Document Id not found.");
             }
-            if (metadataTasks.Count > 0) { await Task.WhenAll(metadataTasks.ToArray()); }
+            existingDocument.DocumentStatusTypeCode = updateRequest.DocumentStatusCode;
+            documentRepository.Update(existingDocument);
 
-            // update the pims document status
-            PimsDocument existingDocument = documentRepository.Get(documentId);
-            if (existingDocument is not null)
+            DocumentUpdateResponse response = new DocumentUpdateResponse()
             {
-                existingDocument.DocumentStatusTypeCode = updateRequest.DocumentStatusCode;
-                documentRepository.Update(existingDocument);
+                MetadataExternalResult = new List<ExternalResult<DocumentMetadata>>(),
+            };
+
+
+            var metadataUpdateSucessfull = false;
+            if (updateRequest.DocumentMetadata.Count > 0)
+            {
+                // Retrieve the existing metadata and check if it needs to be updated.
+                ExternalResult<QueryResult<DocumentMetadata>> existingMetadata = await documentStorageRepository.GetDocumentMetadataAsync(updateRequest.MayanDocumentId);
+
+                List<DocumentMetadataUpdateModel> updates = new List<DocumentMetadataUpdateModel>();
+                List<DocumentMetadataUpdateModel> creates = new List<DocumentMetadataUpdateModel>();
+                List<DocumentMetadataUpdateModel> deletes = new List<DocumentMetadataUpdateModel>();
+
+                foreach (var updateMetadata in updateRequest.DocumentMetadata)
+                {
+                    var existing = existingMetadata.Payload.Results.FirstOrDefault(x => x.MetadataType.Id == updateMetadata.MetadataTypeId);
+
+                    if (existing == null)
+                    {
+                        creates.Add(updateMetadata);
+                    }
+                    else if (existing.Value != updateMetadata.Value && !string.IsNullOrEmpty(updateMetadata.Value))
+                    {
+                        updateMetadata.Id = existing.Id;
+                        updates.Add(updateMetadata);
+                    }
+                    else if (string.IsNullOrEmpty(updateMetadata.Value))
```

Comment:
hmm, just want to double check that this works. So when we want to delete something we send an empty array element from the frontend? Would it be better to check if there are items in the existingMetadata that aren't in update request instead? that would be more in line with the pattern we are using in the repository.

Reply:
I disagree that this does not align with the current pattern we are using. From the form point of view, the way of deleting something is just deleting the text. That does create an entry of the metadata with an empty string. The previous if in this line checks if there is already an element matching in the existing metadata and this if checks if the expected update is the empty string and if so it should be deleted. That leaves the form logic to the minimum necessary.
Changing it so that the front end removes the element if empty would be moving the logic to the frontend, that is what we agreed not to do.